### PR TITLE
modify SecondaryIndexManager#indexPartition() to retrieve only columns for which indexes are actually being built

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe for CASSANDRA-16776

### CODE REVIEW

File: test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java [hunk 1]
```diff
@@ -764,4 +800,135 @@ public void widePartitionsOverlappingRows3() throws Throwable
...
+        cfs.disableAutoCompaction();
+        int rowWidth = 100;
+        int rowsPerPartition = 1000;
+
+        measure(new Workload()
+        {
+            @SuppressWarnings("UnstableApiUsage")
+            public void setup()
+            {
+                cfs.disableAutoCompaction();
```
Comment (azotcsit) [inline test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java:825]: is it necessary to disable automatic compaction twice? it has been already done on line 816.

File: test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java [hunk 2]
```diff
@@ -764,4 +800,135 @@ public void widePartitionsOverlappingRows3() throws Throwable
...
+                String insert = String.format("INSERT INTO %s.%s (k, c, v1, v2, v3, v4) VALUES (?, ?, ?, ?, ?, ?)", ksname, "tbl");
+                for (int f=0; f<numSSTable; f++)
```
Comment (azotcsit) [inline test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java]: nit: missing spaces

File: test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java [hunk 3]
```diff
@@ -764,4 +800,135 @@ public void widePartitionsOverlappingRows3() throws Throwable
...
+                {
+                    for (int p = 0; p < sstablePartitions; p++)
+                    {
+                        String key = String.format("%08d", (f * sstablePartitions) + p);
+                        for (int r=0; r<rowsPerPartition; r++)
```
Comment (azotcsit) [inline test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java]: nit: missing spaces

File: test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java (hunk 1 above)
Comment (maedhroz) [reply test/memory/org/apache/cassandra/db/compaction/CompactionAllocationTest.java:825]: CC @bdeggleston

I think this was a copy/paste, so I'm wondering if it was necessary to disable twice in the original tests like `testTinyPartitions()` et al

File: src/java/org/apache/cassandra/index/internal/CollatedViewIndexBuilder.java [hunk 4]
```diff
@@ -55,29 +57,57 @@ public CollatedViewIndexBuilder(ColumnFamilyStore cfs, Set<Index> indexers, Redu
...
+            {
+                if (index.dependsOn(column))
+                {
+                    builder.add(column);
+                    isPartitionIndex = false;
+                }
+
+                // if any index declares no dependency on any column, it is a full partition index
+                // so we can use the base partition columns as the input source
+                if (isPartitionIndex)
```
Comment (belliottsmith) [inline src/java/org/apache/cassandra/index/internal/CollatedViewIndexBuilder.java]: Shouldn't this be outside of the loop?

Comment (maedhroz) [reply src/java/org/apache/cassandra/index/internal/CollatedViewIndexBuilder.java]: You're right. Good catch :)

Comment (maedhroz) [reply src/java/org/apache/cassandra/index/internal/CollatedViewIndexBuilder.java]: Fixed.

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-16776

**Pull request closed without merging**